Initial benchmarks (#76)
serverless/console: node/packages/aws-lambda-otel-extension/test/benchmark/README.md

# Benchmarks

## Tested use cases

### Node.js function ([`callback`](../../fixtures/lambdas/callback.js))

No-op Node.js function. Handler takes callback and immediately invokes it with dummy payload

### Node.js + Express function ([`express`](../../fixtures/lambdas/express.js))

Simple express app with two endpoints which respond with dummy payload.

> Function is invoked with payload that reflecxts one that function will receive if it would be triggered through API Gateway integration

Reviewer: ```suggestion
Function is invoked with payload that reflects one that function will receive if it would be triggered through API Gateway integration
```

In this case, produced extension invocation overheads are additionally reduced. Tests show that for longer going functions overheads remain same as for that one

### Node.js heavy logging function ([`logger`](../../fixtures/lambdas/logger.js))

Logs 1000 individual log messages via `console.log` in short time frames

It's the only case which writes any logs (in other cases log reporting is not existing).

## Benchmark variants

### `bare`

No instrumentation (bare run with no extension layers onboard).

As there are no extensions loaded, there's no extension duration overhead introduced, and benchmark will naturally report them with zeros.

What we get is metrics from AWS report, that confirm on initialization duration, invocation duration, billed invocation duration and max memory used.

### `externalOnly`

Just external extension which processes lambda event cycle. As no internal extension is loaded, no reports are processed and send by the external extension.

Difference from `bare` is that external extension process is initialized by AWS, and its that process that monitors and queues lambda invocation event cycle.

> This additional process managment introduces natural duration overhead which is best calculated by reading duration as reported by AWS for this variant and subtracting from it a `bare` variant duration _(current numbers show ca 100ms overhead for initialization, ca 60ms for first invocation , and ca 30ms for every following invocation)_

Reviewer: ```suggestion
This additional process management introduces natural duration overhead which is best calculated by reading duration as reported by AWS for this variant and subtracting from it a `bare` variant duration _(current numbers show ca 100ms overhead for initialization, ca 60ms for first invocation , and ca 30ms for every following invocation)_
```

There's no time spent here on processing reports and propagating them to the ingestion server (as none are to be send from the internal extension). So internally measured invocation duration overhead of external extension is also reported as not existing

### `jsonLog`

External & internal extensions are loaded. Internal extension operates fully, while in external extension from obtained reports just payloads are generated and logged into `stdout` (there's no conversion to Protobuf and not reporting to the remote ingestion server).

Difference from `externalOnly` is purely that internal extension is loaded and fully active and that it sends reports to the external extension.

> While external extension prepares the payloads and logs them to the console (in JSON form) overhead of that is neglible (internal ovehead measurements are expected to show `0`).

Reviewer: ```suggestion
While external extension prepares the payloads and logs them to the console (in JSON form) overhead of that is negligible (internal overhead measurements are expected to show `0`).
```

Comparing its duration overhead against `externalOnly` shows how much overhead do internal extension introduce.

### `protoLog`

Exactly same as `jsonLog` with a difference, that in the external extension JSON payloads are converted to the Protobuf (and it's Protobuf output that's logged to the `stdout`)

Comparing duration differences with `jsonLog` we can see how much overhead is generated by Protobuf conversion

### `protoConsole`

Exactly same as `protoLog` but external extension reports the generated payloads to the ingestion server.

Comparing duration differences with `protoLog` we can see how much overhead is generated by communication with the remote ingestion server

## Reported metrics

Each benchmark variant is run against 5 lambda instances.

Therefore for each metric outlined below, we get 5 numbers. Durations returned in result report are averages of those 5 values.

Internally measured durations are written by the extensions to the `stdout`, and are read (together with invocation reports as written by AWS) from CloudWatch logs.

### Observed AWS Lambda execution lifecycle phases

> #### Ininitialization (`init`)

Reviewer: ```suggestion
#### Initialization (`init`)
```

> Lambda instance initilization

Reviewer: ```suggestion
Lambda instance initialization
```

_Currently through benchmark reports we can see that external extension initialization takes ca 100ms, while internally measured initialization takes ca 30ms)_

###### `init:internal`

Duration of the internal extension initialization logic, measured internally in internal extension. It's a time measured from start of the internal extension logic, until it returns for further evaluation by AWS runtime engine.

This initialization is synchronous, but internal extension may propagate some async tasks, of which processing is then mixed with tasks as issued eventually by lambda initialization logic (if it's asynchronous) or first invocation logic.

Due to how otel libraries are designed, internal extension issues an asynchronous task, still it's very minimal and shouldn't introduce any notable overhead.

As internal extension runs in very same process as lambda, there's we don't deal with other process initialization overhead. Still as this extension is handled by AWS runtime logic which otherwise is not triggered, and pollutes process memory with its objects, we still can observe some extra overhead not covered by this metric.

> _Currently through benchmark reports, we can see that internal extension initialization (for noop lambda) takes ca 150ms, while internally measured iintialization takes ca 120ms_

Reviewer: ```suggestion
_Currently through benchmark reports, we can see that internal extension initialization (for noop lambda) takes ca 150ms, while internally measured initialization takes ca 120ms_
```

###### `[first|following]:internal:request`

Invocation request handing duration in context of the Node.js internal extension. It's a time that's measured from the moment when our handler wrapper received an invocation event from AWS until we invoke the original Lambda handler.

It's a synchronous task, but extension logic starts asynchronous request to the external extension, which is processed when original lambda logic is run (or after it ends processing, in case of fast resolving lambda). This means that the extension and original handler tasks ordering is mixed, and we cannot reliably measure internally an overhead as introduced by the internal extension request handling in total.

###### `[first|following]:internal:response`

Invocation response handling duration in the context of the Node.js internal extension. It's a time that's measured from the moment when original handler provided response until we pass it back to AWS.

This one is asynchronous, as we propagate the response only after we're certain that extension request and response handlers have finalized its tasks (send successfully the telemetry data to the external extension).

###### `[first|following]:external`

> Time spent in context of external extension process between recieving the `platform.runtimeDone` event (which signals that ongoing invocation have ended) and reporting via AWS Lambda extension API that we're ready to process next invocation )(external extension reports that only after all started ingestion server requests are finalized)

Reviewer: ```suggestion
Time spent in context of external extension process between receiving the `platform.runtimeDone` event (which signals that ongoing invocation have ended) and reporting via AWS Lambda extension API that we're ready to process next invocation )(external extension reports that only after all started ingestion server requests are finalized)
```

###### `[first|following]:total`

Total duration time as reported by AWS (with `duration` metric provided in invocation report)

###### `[first|following]:billed`

Billed duration time as reported by AWS (with `billedDuration` metric provided in invocation report)

###### `[first|following]:local

Locally measured SDK invocation time. This value is largely influenced by the location from which we run benchmarks and quality of the internet connection.

For benchmarks run against `us-east-1` region, shortest values we'll get e.g. in New York, but running benchmarks in San Francisco will already add extra latency, and numbers get additionally worse, when running benchmarks from farther locations

###### `[first|following]:maxMemoryUsed

Max memory used as reported by AWS (with `maxMemoryUsed` metric provided in invocation report). It's to expose what memory overhead is introduced by the extensions

### Summary

> While overhead durations as measured and reported by the extensions internally help to coin where exactly performance bottlenecks occur, the real extension overhead numbers can only be deducated from durations as reported by the AWS.

Reviewer: ```suggestion
While overhead durations as measured and reported by the extensions internally help to coin where exactly performance bottlenecks occur, the real extension overhead numbers can only be deducted from durations as reported by the AWS.
```

With benchmark variants, as setup correctly. It's safe to assume that:

- Comparing durations of `externalOnly` vs `bare` expose the overhead of the AWS Lambda event lifecycle handling in the external extension
- Comparing durations of `jsonLog` vs `externalOnly` expose the overhead of the internal extension

> - Comparing durations of `protoLog` vs `jsonLog` expose the overhead of the Protobuf conversion that happens in external exteneion

Reviewer: ```suggestion
- Comparing durations of `protoLog` vs `jsonLog` expose the overhead of the Protobuf conversion that happens in external extension
```

- Comparing durations of `protoConsole` vs `protoBuf` expose the overhead of the communication with ingestion server that happens in external extension

## Setup & Run

### 1. Ensure all dependencies are installed:

Run `npm install` in following folders:

- `external/otel-extension-external`
- `internal/otel-extension-internal-node`
- `test/fixtures/lambdas`

### 2. Ensure AWS credentials

### 3. Configure environment variables

- `AWS_REGION` - region against which benchmarks should be run
- `SLS_ORG_NAME` & `SLS_ORG_TOKEN`- (optional) Needed to benchmark scenario of reporting to the Console Kinesis server (if not provided, `protoConsole` variant will not be tested)

> - `TEST_UID` - (optional) common name token to be used as part of genered resource names. All generated resource names will be prefixed with `test-oext-<test-uid>`. If not provided, one is generated on basis of [local machine id](https://www.npmjs.com/package/node-machine-id). Note: Script ensures that all generated resources are removed after benchmark is done

Reviewer: ```suggestion
- `TEST_UID` - (optional) common name token to be used as part of generated resource names. All generated resource names will be prefixed with `test-oext-<test-uid>`. If not provided, one is generated on basis of [local machine id](https://www.npmjs.com/package/node-machine-id). Note: Script ensures that all generated resources are removed after benchmark is done
```
